cdap-common/bin/functions.sh

```diff
@@ -0,0 +1,1085 @@
+    fi
+  fi
+  logecho "$(date) Starting CDAP ${__name} service on ${HOSTNAME}"
+  "${JAVA}" -version 2>>${__logfile}
+  ulimit -a >>${__logfile}
+  __defines+=" ${OPTS}"
+  echo "Running: ${JAVA} ${__defines} -cp ${CLASSPATH} ${MAIN_CLASS} ${MAIN_CLASS_ARGS} ${@}" >>${__logfile}
+  # Start our JVM
+  nohup nice -n ${NICENESS} "${JAVA}" ${__defines} -cp ${CLASSPATH} ${MAIN_CLASS} ${MAIN_CLASS_ARGS} ${@} </dev/null >>${__logfile} 2>&1 &
+  echo $! >${__pidfile}
+  sleep 2 # Now, wait for JVM spinup
+  kill -0 $(<${__pidfile}) >/dev/null 2>&1
+  return $?
+}
+
+#
+# cdap_run_class <class> [arguments]
+# Executes a given class' main method with the CLASSPATH and environment setup
+#
+cdap_run_class() {
+  local readonly __class=${1}
+  shift
+  local readonly __args=${@}
+  local readonly __ret
+  local JAVA_HEAPMAX=${JAVA_HEAPMAX:--Xmx1024m}
+  [[ -z ${__class} ]] && echo "[ERROR] No class name given!" && die "Usage: ${0} run <fully-qualified-class> [arguments]"
+  # Check and set classpath if in development environment. 
+  cdap_check_and_set_classpath_for_dev_environment "${CDAP_HOME}"
+  # Setup classpaths.
+  cdap_set_classpath "${CDAP_HOME}"/master "${CDAP_CONF}"
+  # Setup Java
+  cdap_set_java || return 1
+  cdap_set_spark || logecho "WARN: Could not determine SPARK_HOME! Spark support unavailable!"
+  cdap_set_hive_classpath || return 1
+  # Add proper HBase compatibility to CLASSPATH
+  cdap_set_hbase || exit 1
+  cdap_check_or_create_master_local_dir || die "Could not create local directory"
+  if [[ -n ${__args} ]] && [[ ${__args} != '' ]]; then
+    echo "$(date) Running class ${__class} with arguments: ${__args}"
+  else
+    echo "$(date) Running class ${__class}"
+  fi
+  "${JAVA}" ${JAVA_HEAPMAX} -Dhive.classpath=${HIVE_CLASSPATH} -Duser.dir=${LOCAL_DIR} -Djava.io.tmpdir=${TEMP_DIR} ${OPTS} -cp ${CLASSPATH} ${__class} ${__args}
+  __ret=${?}
+  return ${__ret}
+}
+
+#
+# cdap_check_and_set_classpath_for_dev_environment <home-dir>
+# check and set classpath if in development enviroment
+#
+cdap_check_and_set_classpath_for_dev_environment() {
+  local readonly __home=${1}
+
+  # Detect if we are in development.
+  IN_DEV_ENVIRONMENT=${IN_DEV_ENVIRONMENT:-false}
+
+  # for developers only, add flow and flow related stuff to class path.
+  if [[ ${IN_DEV_ENVIRONMENT} == true ]]; then
+    echo "Constructing classpath for development environment ..."
+    [[ -f "${__home}"/build/generated-classpath ]] && CLASSPATH+=":$(<${__home}/build/generated-classpath)"
+    [[ -d "${__home}"/build/classes ]] && CLASSPATH+=":${__home}/build/classes/main:${__home}/conf/*"
+    [[ -d "${__home}"/../data-fabric/build/classes ]] && CLASSPATH+=":${__home}/../data-fabric/build/classes/main"
+    [[ -d "${__home}"/../common/build/classes ]] && CLASSPATH+=":${__home}/../common/build/classes/main"
+    [[ -d "${__home}"/../gateway/build/classes ]] && CLASSPATH+=":${__home}/../gateway/build/classes/main"
+    export CLASSPATH
+  fi
+  return 0
+}
+
+# FOO
```

Comment:
Oops... placeholder. It's hard to move around and keep your place between multiple files.